### PR TITLE
[NPU Feature] Qwen3.5 NPU FLA and fused-operator patches

### PR DESCRIPTION
**Pull request by ys2025-AI**
# PR type
- [ ] Bug Fix
- [x] New Feature
- [ ] Document Updates
- [ ] More Models or Datasets Support

# PR information

## Description

This PR ports the Qwen3.5 FLA and Ascend NPU fused-operator patches.

### Motivation

Qwen3.5 introduces hybrid attention layers (`linear_attention` + `full_attention`). The `linear_attention` path relies on `chunk_gated_delta_rule` from the `flash-linear-attention` (FLA) library, which contains CUDA-only Triton kernels. On Ascend NPU, these kernels must be redirected to **MindSpeed Triton** implementations to achieve comparable performance.

Without this patch, Qwen3.5 falls back to the pure PyTorch `torch_chunk_gated_delta_rule`, resulting in **~33% slower training** on NPU.

### Main changes

| File                                       | Change                                                       |
| ------------------------------------------ | ------------------------------------------------------------ |
| `twinkle/kernel/chunk_gated_delta_rule.py` | **New**. MindSpeed Triton wrapper for `chunk_gated_delta_rule`. Re-exports the public API with identical signature to the FLA library. |
| `twinkle/kernel/monkey_patch_npu.py`       | **Extended**. Adds `_patch_qwen3_5_fla()` which: (1) spoofs `transformers.utils.is_flash_linear_attention_available` to bypass CUDA-only checks; (2) replaces module-level `chunk_gated_delta_rule` with the MindSpeed implementation; (3) traverses instantiated model layers to re-bind per-instance `chunk_gated_delta_rule` (required because Qwen3.5 caches the function reference at `__init__` time). |

### Environment Variables

All FLA behavior is gated under the existing `TWINKLE_NPU_PATCH` hierarchy:

```bash
# Master switch for all NPU optimizations.
export TWINKLE_NPU_PATCH=1

# Enable fused operators (RMSNorm, RoPE, SwiGLU, SDPA).
export TWINKLE_NPU_FUSED_OPS=1

# Enable MoE Grouped MatMul.
export TWINKLE_NPU_MOE_PATCH=1

# Enable Flash Linear Attention for Qwen3.5.
# Default: 1 (enabled). Set to 0 to force torch fallback.
export TWINKLE_NPU_FLA=1
```
**Related**: [modelscope/ms-swift#9223](https://github.com/modelscope/ms-swift/pull/9223)
## Experiment results

- **Model**: `Qwen3.5-35B-A3B` (40 layers, 30× `linear_attention`)
- **Hardware**: Atlas 900 A3 (2 x NPU)
- **Dataset**: GSM8K_ZH
- **Finetuning type**: LoRA
- **Software**: cann8.5.0 + torch/orch_npu 2.9.0 + MindSpeed 0.12.1 + triton-ascend 3.2.0 + transformers 5.9

| Metric                                                          | Baseline (FLA OFF) | FLA ON (MindSpeed) | Delta                     |
| --------------------------------------------------------------- | ------------------ | ------------------ | ------------------------- |
| **Avg. duration per 10-step interval**                          | **57.7 s**         | **43.8 s**         | **−24.1%**                |
| **Avg. Loss**                                      |—              |—              | **−0.0005** (identical)   |
| **Avg. Grad Norm**                                 |—               |—               | **−0.024** (within noise) |

### CODE REVIEW

File: src/twinkle/kernel/chunk_gated_delta_rule.py [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+            q, k, v, g, beta,
+            initial_state=h0,
+            output_final_state=True,
+            cu_seqlens=cu_seqlens
+        )
+    """
+    import torch.distributed as dist
+    rank = dist.get_rank() if dist.is_initialized() else 0
+    #if rank == 0:
+    #    print(f"[FLA] chunk_gated_delta_rule called, q.shape={q.shape}, k.shape={k.shape}, v.shape={v.shape}")
```
[the remark this answers is not in the document]
Comment (ys2025-AI) [reply src/twinkle/kernel/chunk_gated_delta_rule.py]: 已经去除

File: src/twinkle/kernel/monkey_patch_npu.py [hunk 2]
```diff
@@ -85,7 +107,137 @@ def apply_hf_moe_grouped_mm_patch() -> None:
...
+    router_indices = router_indices.to(torch.int32)
+
+    permuted_hidden_states, row_ids_map = torch_npu.npu_moe_token_permute(hidden_states, router_indices)
+    tokens_per_expert = torch.histc(
+        router_indices.to(torch.float),
+        bins=self.num_experts,
+        min=0,
+        max=self.num_experts,
+    ).to(torch.int64)
+    gate_up_weight, down_weight = _normalize_packed_expert_weights(self, hidden_states.dtype, hidden_dim)
```
[the remark this answers is not in the document]
Comment (ys2025-AI) [reply src/twinkle/kernel/monkey_patch_npu.py]: 已经修改

File: src/twinkle/kernel/monkey_patch_npu.py [hunk 3]
```diff
@@ -96,33 +248,105 @@ def __init__(self, hidden_size: int, eps: float = 1e-6) -> None:
...
+def npu_gated_rms_norm_forward(self, hidden_states, gate=None):
+    """NPU forward for Gated RMSNorm.
+
+    Keeps the original activation dtype by default to avoid FP32 cast overhead.
+    Set ``TWINKLE_NPU_GATED_RMSNorm_FP32=1`` to force high-precision FP32.
+    """
+    input_dtype = hidden_states.dtype
+    _eps = getattr(self, 'variance_epsilon', getattr(self, 'eps', 1e-6))
+
+    force_fp32 = os.environ.get('TWINKLE_NPU_GATED_RMSNorm_FP32', '0').lower() in ('1', 'true', 'on', 'yes')
```
[the remark this answers is not in the document]
Comment (ys2025-AI) [reply src/twinkle/kernel/monkey_patch_npu.py]: 已经修改

**Review by tastelikefeet — APPROVED** [review]

File: src/twinkle/kernel/monkey_patch_npu.py [hunk 4]
```diff
@@ -85,7 +107,137 @@ def apply_hf_moe_grouped_mm_patch() -> None:
...
+    routing_weights = routing_weights.to(hidden_states.dtype)
+    router_indices = router_indices.to(torch.int32)
+
+    permuted_hidden_states, row_ids_map = torch_npu.npu_moe_token_permute(hidden_states, router_indices)
+    tokens_per_expert = torch.histc(
+        router_indices.to(torch.float),
+        bins=self.num_experts,
+        min=0,
+        max=self.num_experts,
+    ).to(torch.int64)
```
[the remark this answers is not in the document]
Comment (ys2025-AI) [reply src/twinkle/kernel/monkey_patch_npu.py]: 已经修改
